Add blog post for R118 - New bad row format (#1313)
snowplow/snowplowanalytics.com: _posts/2020-01-13-release-r118/index.md

In addition, this release also includes important refactoring in the libraries that we use (e.g. `cats` instead of `scalaz` and `circe` instead of`json4s`),
a bump of [Beam](https://beam.apache.org/) version to 2.11.0,
an [updated](https://snowplowanalytics.com/blog/2019/08/09/iglu-scala-client-0.6.0-released/) [Iglu Client](https://github.com/snowplow/iglu-scala-client)
with different validator and fixed caching mechanism,
as well as an improvement to the referer parser enrichment.

[Scala Stream Collector](https://github.com/snowplow/snowplow/tree/master/2-collectors/scala-stream-collector)
and the enrich jobs versions ([Stream Enrich](https://github.com/snowplow/snowplow/tree/master/3-enrich/stream-enrich) and
[Beam Enrich](https://github.com/snowplow/snowplow/tree/master/3-enrich/beam-enrich)) are bumped to `1.0.0`.

> As introduced in the [RFC](https://discourse.snowplowanalytics.com/t/rfc-making-the-snowplow-pipeline-real-time-end-to-end-and-deprecating-support-for-batch-processing-modules/3018),

Reviewer: ```suggestion
As introduced in the [RFC on batch deprecation](https://discourse.snowplowanalytics.com/t/rfc-making-the-snowplow-pipeline-real-time-end-to-end-and-deprecating-support-for-batch-processing-modules/3018),
```

> we stop supporting Spark from this release, in favor of streaming.

Reviewer: ```suggestion
we are no longer adding new features to Spark from this release. This release, and all subsequent, will focus on our streaming solution.
```
Just added a bit more emphasis to reduce the chances of this being missed. We were at one point going to add the capability to Spark Enrich. Did we back out of this?

1. [Beta release](#betarelease)
2. [New bad row format](#badrows)
3. [Referer parser enrichment improvement](#refererparser)
4. [Upgrading](#upgrading)
5. [Roadmap](#roadmap)
6. [Getting help](#help)

<h1 id="betarelease">1. Beta release</h1>

This release represents a major upgrade of the core Snowplow technology. We are currently testing this tech internally - no small feat because of the size and scope of the release. Whilst we are running those tests, we have decided to make the code available to open source users to test themselves: we would welcome any feedback on the new functionality, which we hope to release in production in [R119](#roadmap).

> If you are an open source users who would like to test this new functionality, we would be happy to provide a high-priority OSS support during this testing phase (covering setup, debug, recovery or other aspects). We are committed to iterating our release candidates for R119 rapidly.

Reviewer: ```suggestion
If you are an open source user who would like to test this new functionality, we would be happy to provide a high-priority OSS support during this testing phase (covering setup, debug, recovery or other aspects). We are committed to iterating our release candidates for R119 rapidly.
```

<h1 id="badrows">2. New bad row format</h1>

<h2>2.1. Why update the bad row format?</h2>

> The Snowplow pipeline has always been non-lossy: when there is an issue processing an event, rather than silently discard the event, it is saved as a "bad row" - opening up the possibility to monitor the number of bad rows, inspect them, fix them and reprocess them.

Reviewer: ```suggestion
The Snowplow pipeline has always been non-lossy: when there is an issue processing an event, rather than silently discard the event, it is saved as a "bad row" . This is only half the challenge! Snowplow also gives the ability to monitor the number of bad rows, inspect them, fix them and reprocess them to complete your data set.
```

The previous format for these bad rows (generic JSON) was not straightforward to use. Specifically:

1. It was hard to distinguish "bad rows" that mattered i.e. that were real data that was not successfully processed, with rubbish data e.g. generated by bots crawling the collector endpoint
2. Where there were geniune issues, it was hard to 
  1. aggregate over the bad row data to size those issues, 
  2. figure out what type of data caused the issue (e.g. for a schema validation error, what schema the issue was with, and what field the issue was with), and
  3. identify the source of the error so that it can be quickly fixed e.g. is it an issue with a specific mobile app?

> In an effort to address these issues, we decided to rethink the format from the ground up. More background on that thinking can be found on the [RFC](https://discourse.snowplowanalytics.com/t/a-new-bad-row-format/2558).

Reviewer: ```suggestion
To address these issues, we have reworked the format from the ground up. More background on that thinking can be found on the [RFC](https://discourse.snowplowanalytics.com/t/a-new-bad-row-format/2558).
```
We've done it!
